File: Greedy/Kruskal’sMinimumSpanningTree/Kruskal’sMinimumSpanningTree.java
import java.util.Arrays;
import java.util.Scanner;

public class KruskalsMinimumSpanningTree {
    int V;
    int E; //Number of vertices and edges in the graph
    Edge[] edge;
    Edge[] mst; //Array of Edge holds the entire graph and mst array holds the Edges that are in the mst
    int[] parent; //Disjoint-set
    int[] size; //Size array for size of set

    public static class Edge implements Comparable<Edge> {
        //beginning vertex, ending vertex, weight of edge
        int bv;
        int ev;
        int cost;

        //Empty constructor
        public Edge() {
            //to initialize arrays later
        }

        //Full constructor
        public Edge(int bv, int ev, int cost) {
            this.bv = bv;
            this.ev = ev;
            this.cost = cost;
        }

        @Override
        public int compareTo(Edge other) {
            return this.cost - other.cost;
        }
    }

    public KruskalsMinimumSpanningTree(int v, int e) {
        V = v;
        E = e;

        edge = new Edge[e];
        for (int i = 0; i < e; i++) {
            edge[i] = new Edge();
        }

        mst = new Edge[v - 1];
        for (int i = 0; i < v - 1; i++) {
            mst[i] = new Edge();
        }

        parent = new int[v];
        for (int i = 0; i < v; i++) {
            parent[i] = -1;
        }

        size = new int[v];
        for (int i = 0; i < v; i++) {
            size[i] = 1;
        }
    }

    public int find(int v) {
        if (parent[v] == -1) { // if -1 it is the root/parent
            return v; //the vertex is already the parent
        } else {
            return find(parent[v]); //if it's not the parent, keep using find to find the parent
        }
    }

    public void union(int bv, int ev) {
        int pb = find(bv); //parent of beginning vertex
        int pe = find(ev); //parent of beginning vertex

        if (size[pb] < size[pe]) { //if the size of one set is greater than the other
            /*
             * set the parent of the smaller set to the parent of the larger set,
             * we're attaching the ENTIRE smaller set from its parent vertex to
             * the parent vertex of the larger set.
             */
            parent[pb] = pe;
            /*
             * add the size of the smaller set to the size of the
             * larger set (since they're 1 set now)
             */
            size[pe] += size[pb];
        } else {
            //same procedure as above
            parent[pe] = pb;
            size[pb] += size[pe];
        }
    }

    public static void main(String[] args) {
        //Create a Scanner so we can input the information about edges
        Scanner sc = new Scanner(System.in);

        //Let's input how many vertices and edges we're given
        int v = sc.nextInt();
        int e = sc.nextInt();

        //Create a new object of your Main class, let's call it "graph"
        //and pass in the parameters v (vertices) and e (edges)
        //The constructor of the Main class will initialize all our
        //arrays for us.
        KruskalsMinimumSpanningTree graph = new KruskalsMinimumSpanningTree(v, e);

        //Using a for-loop, input the information about each edge
        for (int i = 0; i < e; i++) {
            int bv = sc.nextInt(); //beginning vertex
            int ev = sc.nextInt(); //ending vertex
            int cost = sc.nextInt();
            //Now let's use the 2nd constructor of the Edge class
            //and put the above information into our Edge array
            graph.edge[i] = new Edge(bv, ev, cost);
        }

        //Using Arrays.sort(), we make use of the Comparable interface
        //we implemented in the Edge class
        Arrays.sort(graph.edge);

        //Create a count variable to keep track of the edges we've added
        int count = 0;
        //Create a for-loop to loop through all the given edges
        //we sorted earlier
        for (int i = 0; i < e; i++) {
            //Grab the details of the ith edge
            //it should be the edge with the least cost
            int bv = graph.edge[i].bv;
            int ev = graph.edge[i].ev;
            int cost = graph.edge[i].cost;

            //Using the find function we created earlier
            //for our disjoint-set, use it to find bv's root/parent
            //and ev's root/parent. Store in respective variables
            int pb = graph.find(bv); //parent of beginning vertex
            int pe = graph.find(ev); //parent of ending vertex

            //If the parent of bv and ev are not the same,
            //then the edge won't form a cycle
            if (pb != pe) {
                //Using the union function
                graph.union(bv, ev);

                //Add the edge to the MST array
                //Using count because not every given edge (i)
                //can be added to the MST
                graph.mst[count].bv = bv;
                graph.mst[count].ev = ev;
                graph.mst[count].cost = cost;

                //If the MST has V - 1 edges in it
                //then we have found the MST of the graph
                //WE'VE COMPLETED THE ALGORITHM!
                if (count == v - 1) {
                    break;
                }
            }
        }
        for (int i = 0; i < v - 1; i++) {
            System.out.print(graph.mst[i].bv + " ");
            System.out.print(graph.mst[i].ev + " ");
            System.out.println(graph.mst[i].cost);
        }
    }
}

File: Greedy/Kruskal’sMinimumSpanningTree/Kruskal’sMinimumSpanningTree.py
import math, random
import numpy as np

def find(b, c):
	for a in visited:
		if (b in a) and (c in a):
			return False
		else:
			return True

n = random.randint(5,7)
print('Number of vertices:', n)
max_int = math.inf
num, m, l = 0, 0, 0

e = [[0 for x in range(n)] for y in range(n)] 
c = [[max_int for x in range(n)] for y in range(n)]
t = [[0 for x in range(2)] for y in range(n-1)] 4

print (np.matrix(t)) 

for i in range(n):
	for j in range(n):
		if(i < j):
			e[i][j]=1
			c[i][j]=float(random.randint(10,50))
		else:
			e[i][j]=e[j][i]
			c[i][j]=c[j][i]

print('Edge Adjacency Matrix: \n', np.matrix(e))
print('Cost Adjacency Matrix: \n',np.matrix(c))
mincost = 0
visited = [set() for _ in range(n)]
new_set = set()
for i in visited:
	i.add(num)
	num += 1
print ('Disjoint Sets: ', visited)
values = np.array(c)
count = 0
while(count < n-1):
	index = np.argmin(values)
	u, v = index//n, index%n
	if find(u, v):
		t[count][0], t[count][1] = u, v
		values[u][v], values[v][u] = max_int, max_int
		for a in visited:
			if u in a:
				k = a	
			if v in a:
				l = a
		visited.append(k.union(l))
		visited.remove(k)
		visited.remove(l)		
		mincost += c[u][v]
		count += 1
if (count != n-1):
	print ('No MST was formed.')

print ('MST: \n', np.matrix(t))
print ('Visited:', visited)
print ('Cost:', mincost)

		
